src/load_cell_endstop.c

```diff
@@ -0,0 +1,297 @@
+// Load Cell based end stops.
+//
+// Copyright (C) 2023  Gareth Farrington <gareth@waves.ky>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include "basecmd.h" // oid_alloc
+#include "board/irq.h" // irq_disable
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // shutdown
+#include "trsync.h" // trsync_do_trigger
+#include "board/misc.h" // timer_read_time
+#include "sos_filter.h" // fixedQ12_t
+#include "load_cell_endstop.h" //load_cell_endstop_report_sample
+#include <stdint.h> // int32_t
+#include <stdlib.h> // abs
+
+// Q33.30
+typedef int64_t fixedQ33_t; // Q1.30 value stored in int64
+#define FIXEDQ33_FRAC_BITS FIXEDQ1_FRAC_BITS
+#define MAX_TRIGGER_GRAMS ((1 << FIXEDQ16) - 1)
+#define FIXEDQ1_ONE 1 << FIXEDQ1_FRAC_BITS
+
+// Flags
+enum {FLAG_IS_HOMING = 1 << 0
+    , FLAG_IS_TRIGGERED = 1 << 1
+    , FLAG_IS_HOMING_TRIGGER = 1 << 2
+    };
+
+// Endstop Structure
+struct load_cell_endstop {
+    struct timer time;
+    uint32_t trigger_grams, trigger_ticks, last_sample_ticks, rest_ticks;
+    struct trsync *ts;
+    int32_t last_sample, safety_counts_min, safety_counts_max, tare_counts;
+    uint8_t flags, trigger_reason, error_reason, watchdog_max
+            , watchdog_count;
+    fixedQ16_t trigger_grams_fixed;
+    fixedQ1_t grams_per_count;
+    struct sos_filter *sf;
+};
+
+// returns the integer part of a fixedQ48_t
+static inline int64_t
+round_fixedQ48(const fixedQ48_t fixed_value) {
+    return fixed_value >> FIXEDQ48_FRAC_BITS;
+}
+
+// Convert sensor counts to grams
+static inline fixedQ48_t
+counts_to_grams(struct load_cell_endstop *lce, const int32_t counts) {
+    // tearing ensures readings are referenced to 0.0g
+    const int32_t delta = counts - lce->tare_counts;
+    // convert sensor counts to grams by multiplication: 124 * 0.051 = 6.324
+    // this optimizes to single cycle SMULL instruction
+    const fixedQ33_t product = (int64_t)delta * (int64_t)lce->grams_per_count;
+    // after multiplication there are 30 fraction bits, reduce to 15
+    // caller verifies this wont overflow a 32bit int when truncated
+    const fixedQ48_t grams = product >>
+                                (FIXEDQ33_FRAC_BITS - FIXEDQ48_FRAC_BITS);
+    return grams;
+}
+
+static inline uint8_t
+is_flag_set(const uint8_t mask, struct load_cell_endstop *lce)
+{
+    return !!(mask & lce->flags);
+}
+
+static inline void
+set_flag(uint8_t mask, struct load_cell_endstop *lce)
+{
+    lce->flags |= mask;
+}
+
+static inline void
+clear_flag(uint8_t mask, struct load_cell_endstop *lce)
+{
+    lce->flags &= ~mask;
+}
```

Comment:
Not a blocker, but for what it is worth, I find these wrappers make the code harder to understand.

Reply:
I find the opposite. Maybe its because I'm not a C native. I have to stop and reason about each one of the bit twiddling statements every time. `is_flag_set(FLAG_IS_HOMING_TRIGGER, lce)` is much faster to read for me. Its a lot closer to reading like `flag is True` in python.